File: client_examples/xdcc_download_client.js
#!/usr/bin/env node

/**
 * XDCC Download Client
 *
 * A Node.js client for the XDCC Download Server
 */

const net = require('net');
const readline = require('readline');

// Parse command line arguments
const args = process.argv.slice(2);
let host = 'localhost';
let port = 8080;
let botName = '';
let packNumber = '';
let sendProgress = true;

// Parse arguments
for (let i = 0; i < args.length; i++) {
  if (args[i] === '--host' && i+1 < args.length) {
    host = args[i+1];
    i++;
  } else if (args[i] === '--port' && i+1 < args.length) {
    port = parseInt(args[i+1], 10);
    i++;
  } else if (args[i] === '--bot' && i+1 < args.length) {
    botName = args[i+1];
    i++;
  } else if (args[i] === '--pack' && i+1 < args.length) {
    packNumber = args[i+1];
    i++;
  } else if (args[i] === '--no-progress') {
    sendProgress = false;
  } else if (args[i] === '--help') {
    console.log(`
XDCC Download Client

Usage:
  node client.js --host <host> --port <port> --bot <bot_name> --pack <pack_number> [--no-progress]

Options:
  --host <host>       Server hostname (default: localhost)
  --port <port>       Server port (default: 8080)
  --bot <bot_name>    Bot name (required)
  --pack <pack_number> Pack number (required)
  --no-progress       Don't request progress updates
  --help              Show this help message
`);
    process.exit(0);
  }
}

// Validate required arguments
if (!botName || !packNumber) {
  console.error('Error: Bot name and pack number are required');
  console.error('Use --help for usage information');
  process.exit(1);
}

// Format bytes to human-readable size
function formatSize(bytes) {
  if (bytes === 0) return '0 Bytes';
  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
}

// Draw a progress bar
function drawProgressBar(percent, width = 30) {
  const filled = Math.floor(width * percent / 100);
  const bar = '█'.repeat(filled) + '-'.repeat(width - filled);
  return `[${bar}] ${percent}%`;
}

// Create a socket connection
const socket = new net.Socket();
const startTime = Date.now();
let lastProgress = 0;
let buffer = '';

// Connect to the server
console.log(`Connecting to ${host}:${port}...`);
socket.connect(port, host, () => {
  console.log('Connected to XDCC Download Server');

  // Create the request
  const request = {
    bot_name: botName,
    pack_number: packNumber,
    send_progress: sendProgress
  };

  console.log(`Sending request: ${JSON.stringify(request)}`);
  socket.write(JSON.stringify(request));
});

// Handle data received from the server
socket.on('data', (data) => {
  // Add the data to our buffer
  buffer += data.toString();

  // Process complete JSON objects
  let startPos = 0;
  while ((startPos = buffer.indexOf('{', startPos)) !== -1) {
    try {
      // Find a potential JSON object
      let endPos = buffer.indexOf('}', startPos);
      if (endPos === -1) break;

      // Extract and parse the JSON
      const jsonStr = buffer.substring(startPos, endPos + 1);
      const response = JSON.parse(jsonStr);

      // Remove the processed part from the buffer
      buffer = buffer.substring(endPos + 1);
      startPos = 0;

      // Process the response based on its status
      switch (response.status) {
        case 'downloading':
          console.log(`Server accepted request: ${response.message}`);
          break;

        case 'progress':
          const progress = response.progress;
          const received = response.received;
          const total = response.total;

          // Update tracking variables
          lastProgress = progress;

          // Calculate speed
          const elapsed = (Date.now() - startTime) / 1000; // seconds
          const speed = received / elapsed;

          // Print progress bar
          const progressBar = drawProgressBar(progress);
          process.stdout.write(`\r${progressBar} | ${formatSize(received)}/${formatSize(total)} | ${formatSize(speed)}/s`);
          break;

        case 'success':
          console.log(`\nDownload completed successfully!`);
          console.log(`File: ${response.filename}`);
          console.log(`Size: ${formatSize(response.size)}`);
          console.log(`Saved to: ${response.path}`);
          socket.end();
          break;

        case 'error':
          console.log(`\nDownload failed: ${response.message}`);
          socket.end();
          break;

        default:
          console.log(`\nUnknown status: ${response.status}`);
          console.log(response);
      }
    } catch (err) {
      // If we couldn't parse it as JSON, move past this opening brace
      startPos++;
    }
  }
});

// Handle socket closure
socket.on('close', () => {
  if (lastProgress > 0) {
    console.log(`\nServer closed connection after reaching ${lastProgress}% progress.`);
    console.log(`The download is likely continuing on the server.`);
    if (lastProgress > 90) {
      console.log(`Download was at ${lastProgress}%, likely completed successfully!`);
    } else {
      console.log(`Download may still be in progress on the server.`);
    }
  } else {
    console.log('\nConnection closed');
  }
});

// Handle errors
socket.on('error', (err) => {
  console.error(`Error: ${err.message}`);
  process.exit(1);
});

// Handle user interruption
process.on('SIGINT', () => {
  console.log('\nOperation cancelled by user');
  socket.end();
  process.exit(0);
});

File: client_examples/xdcc_download_client.py
#!/usr/bin/env python3
"""
XDCC Download Client

A simple client to request files from the XDCC Download Server.
"""

import socket
import json
import argparse
import time
import sys


def format_size(bytes_size):
    """Format file size to human-readable format"""
    if bytes_size < 1024:
        return f"{bytes_size} B"
    elif bytes_size < 1024 * 1024:
        return f"{bytes_size / 1024:.2f} KB"
    elif bytes_size < 1024 * 1024 * 1024:
        return f"{bytes_size / (1024 * 1024):.2f} MB"
    else:
        return f"{bytes_size / (1024 * 1024 * 1024):.2f} GB"


def draw_progress_bar(percent, width=30):
    """Draw a text-based progress bar"""
    filled = int(width * percent / 100)
    bar = '█' * filled + '-' * (width - filled)
    return f"[{bar}] {percent}%"


def main():
    """Main function to run the client"""
    parser = argparse.ArgumentParser(description="XDCC Download Client")
    parser.add_argument("--host", default="localhost", help="XDCC server host")
    parser.add_argument("--port", type=int, default=8080, help="XDCC server port")
    parser.add_argument("--bot", required=True, help="Bot name (e.g., 'BotName')")
    parser.add_argument("--pack", required=True, help="Pack number")
    parser.add_argument("--no-progress", action="store_true", help="Don't request progress updates")

    args = parser.parse_args()

    # Create socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    try:
        # Connect to server
        print(f"Connecting to {args.host}:{args.port}...")
        sock.connect((args.host, args.port))
        print(f"Connected to XDCC Download Server")

        # Create and send request
        request = {
            "bot_name": args.bot,
            "pack_number": args.pack,
            "send_progress": not args.no_progress
        }

        print(f"Sending request: {json.dumps(request)}")
        sock.sendall(json.dumps(request).encode('utf-8'))

        # Set timeout for receiving data
        sock.settimeout(60)  # 60 seconds

        # Buffer for incomplete data
        buffer = b""

        # Start time for progress calculation
        start_time = time.time()

        print("\nWaiting for server response...")

        # Variables to track download
        last_progress = 0

        # Main receiving loop
        while True:
            try:
                # Try to receive data
                chunk = sock.recv(4096)

                if not chunk:
                    if last_progress > 0:
                        print(f"\nServer closed connection after reaching {last_progress}% progress.")
                        print(f"The download is likely continuing on the server.")
                        if last_progress > 90:
                            print(f"Download was at {last_progress}%, likely completed successfully!")
                        else:
                            print(f"Download may still be in progress on the server.")
                        return 0
                    else:
                        print("\nServer closed connection without any progress updates.")
                        return 1

                # Add to our buffer
                buffer += chunk

                # Try to process one or more JSON objects from buffer
                while True:
                    # Find the start of a JSON object
                    start = buffer.find(b'{')
                    if start == -1:
                        break

                    # Find the end of the JSON object
                    end = buffer.find(b'}', start)
                    if end == -1:
                        break

                    # Extract potential JSON data
                    json_data = buffer[start:end + 1]

                    try:
                        # Try to parse it as JSON
                        response = json.loads(json_data)

                        # If successful, remove from buffer
                        buffer = buffer[end + 1:]

                        # Process the response
                        status = response.get("status", "unknown")

                        if status == "downloading":
                            print(f"Server accepted request: {response.get('message', '')}")

                        elif status == "progress":
                            # Update progress information
                            progress = response.get("progress", 0)
                            filename = response.get("filename", "unknown")
                            received = response.get("received", 0)
                            total = response.get("total", 0)

                            # Update tracking variables
                            last_progress = progress

                            # Calculate speed
                            elapsed = time.time() - start_time
                            speed = received / elapsed if elapsed > 0 else 0

                            # Print progress bar
                            progress_bar = draw_progress_bar(progress)
                            progress_msg = (f"\r{progress_bar} | "
                                            f"{format_size(received)}/{format_size(total)} | "
                                            f"{format_size(speed)}/s")
                            sys.stdout.write(progress_msg)
                            sys.stdout.flush()

                        elif status == "success":
                            print(f"\nDownload completed successfully!")
                            print(f"File: {response.get('filename', 'unknown')}")
                            print(f"Size: {format_size(response.get('size', 0))}")
                            print(f"Saved to: {response.get('path', 'unknown')}")
                            return 0

                        elif status == "error":
                            print(f"\nDownload failed: {response.get('message', 'Unknown error')}")
                            return 1

                    except json.JSONDecodeError:
                        # Not valid JSON - might be incomplete or not JSON at all
                        # Just move past the opening brace and try again
                        buffer = buffer[start + 1:]

            except socket.timeout:
                print("\nTimeout waiting for server response")
                if last_progress > 0:
                    print(f"Download reached {last_progress}% and may be continuing on the server")
                return 1

            except KeyboardInterrupt:
                print("\nOperation cancelled by user")
                return 1

            except Exception as e:
                print(f"\nUnexpected error: {str(e)}")
                return 1

    finally:
        print("\nClosing connection...")
        try:
            sock.close()
        except:
            pass


if __name__ == "__main__":
    sys.exit(main())